src/controllers/containerapp.go

```diff
@@ -0,0 +1,230 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dana-team/container-app-operator/api/v1alpha1"
+	"github.com/dana-team/platform-backend/src/types"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ContainerAppController interface {
+	// CreateContainerApp creates a new container app in the specified namespace.
+	CreateContainerApp(namespace string, capp types.Capp) (types.Capp, error)
+
+	// GetContainerApps gets all container apps from a specific namespace.
+	GetContainerApps(namespace string, cappQuery types.CappQuery) (types.CappList, error)
+
+	// GetContainerApp gets a specific container app from the specified namespace.
+	GetContainerApp(namespace, name string) (types.Capp, error)
+
+	// PatchContainerApp patches a specific container app in the specified namespace.
+	PatchContainerApp(namespace, name string, capp types.Capp) (types.Capp, error)
+
+	// DeleteContainerApp deletes a specific container app in the specified namespace.
+	DeleteContainerApp(namespace, name string) (types.CappError, error)
+}
+
+type containerAppController struct {
+	client client.Client
+	ctx    context.Context
+	logger *zap.Logger
+}
+
+func NewContainerAppController(client client.Client, context context.Context, logger *zap.Logger) (ContainerAppController, error) {
+	return &containerAppController{
+		client: client,
+		ctx:    context,
+		logger: logger,
+	}, nil
+}
+
+func (c *containerAppController) CreateContainerApp(namespace string, capp types.Capp) (types.Capp, error) {
+	c.logger.Debug(fmt.Sprintf("Trying to create capp in namespace: %q", namespace))
+
+	newCapp := createV1alpha1Capp(capp)
+	err := c.client.Create(c.ctx, &newCapp)
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("Could not create capp in namespace %q with error: %v", namespace, err.Error()))
+		return types.Capp{}, err
+	}
+
+	return capp, nil
+}
+
+func (c *containerAppController) GetContainerApps(namespace string, cappQuery types.CappQuery) (types.CappList, error) {
+	c.logger.Debug(fmt.Sprintf("Trying to fetch all capps in namespace: %q", namespace))
+
+	cappList := &v1alpha1.CappList{}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: convertLabelsToMap(cappQuery.Labels),
+	})
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("Could not create label selector with error: %v", err.Error()))
+		return types.CappList{}, err
+	}
+
+	err = c.client.List(c.ctx, cappList, &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: selector,
+	})
+	if err != nil {
+		c.logger.Error(fmt.Sprintf("Could not fetch capps in namespace %q with error: %v", namespace, err.Error()))
+		return types.CappList{}, err
+	}
```

Comment:
Should we switch it to this? It's more idiomatic but maybe less readable

```go
	if err := c.client.List(c.ctx, cappList, &client.ListOptions{Namespace: namespace, LabelSelector: selector}); err != nil {
		c.logger.Error(fmt.Sprintf("Could not fetch capps in namespace %q with error: %v", namespace, err.Error()))
		return types.CappList{}, err
	}
```

Reply:
I disagree, because we'll get very long line which is hard to read in smaller screens (and it's longest the line in the script, more than 125 characters).

Reply:
Yeah no problem, it's fine like this as well